File: JAVA/Brute Force/_121_BestTimeToBuyAndSellStock.java
// Problem: https://leetcode.com/problems/best-time-to-buy-and-sell-stock/

/* Solution: (Brute Force)
 * 1. Iterate through the array
 * 2. Keep track of the minimum price
 * 3. Calculate the profit
 * 4. Update the maximum profit
 * 5. Return the maximum profit
 */

public class _121_BestTimeToBuyAndSellStock {
    private static int maxProfit(int[] prices){
        int maxProfit = 0;
        for(int i = 0; i < prices.length; i++){
            for(int j = i+1; j < prices.length; j++){
                if(prices[j] > prices[i]){
                    maxProfit = Math.max(maxProfit, prices[j] - prices[i]);
                }
            }
        }
        return maxProfit;
    }
    public static void main(String[] args) {
        int[] prices = {7,1,5,3,6,4};
        System.out.println(maxProfit(prices));
    }
}

// Time Complexity: O(n^2)
// Space Complexity: O(1)

File: C++/_121_BestTimeToBuyAndSellStock.cpp
// ✔️ Arrays

// Problem: https://leetcode.com/problems/best-time-to-buy-and-sell-stock/

/* Solution: (Optimal Solution)
 * 1. Iterate through the array
 * 2. Keep track of the minimum price
 * 3. Calculate the profit
 * 4. Update the maximum profit
 * 5. Return the maximum profit
 */

#include <bits/stdc++.h>
#include <vector>
using namespace std;

int maxProfit(vector<int> prices){
    int maxProfit = 0;
    int minPrice = prices[0];
    for(int i = 1; i < prices.size(); i++){
        int cost = prices[i] - minPrice;
        maxProfit = max(maxProfit, cost);
        minPrice = min(minPrice, prices[i]);
    }
    return maxProfit;
}

int main() {
    vector<int> prices = {7, 1, 5, 3, 6, 4};
    cout << maxProfit(prices) << endl;
    return 0;
}

// Time Complexity: O(n)
// Space Complexity: O(1)